cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/store/AppMetadataStore.java

```diff
@@ -177,7 +208,17 @@ public void updateAppSpec(String namespaceId, String appId, String versionId, Ap
     MDSKey key = getProgramKeyBuilder(TYPE_WORKFLOW_NODE_STATE, workflowRunId.getParent())
       .add(workflowRunId.getRun()).build();
 
-    return list(key, WorkflowNodeStateDetail.class);
+    List<WorkflowNodeStateDetail> nodeStateDetails = list(key, WorkflowNodeStateDetail.class);
+
+    // Check again without the version to account for old data format since they might not have been updated yet
+    // Since all the programs needs to be stopped before upgrade, either we will have node state details for
```

Comment:
I'm not sure I understand this comment - `Since all the programs needs to be stopped before upgrade`. Does upgrade here mean CDAP package upgrade?